### ISSUE TITLE
Crate does not compile with newly released ash 0.33

### ISSUE THREAD

**Issue by aclysma** [gwihlidal/vk-mem-rs#54]
First, thanks for making these bindings!

I'm hitting this compile error since ash 0.33 released. (I haven't looked into the correct fix yet, but wanted to mention it)

```
error[E0432]: unresolved import `ash::version`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/vk-mem-0.2.2/src/lib.rs:753:18
    |
753 |         use ash::version::{DeviceV1_0, DeviceV1_1, InstanceV1_0};
    |                  ^^^^^^^ could not find `version` in `ash`
```

Also wanted to mention, since vk-mem does not use semver to reference ash (cargo.toml has ash >= 0.27.1), the new ash release broke CI in my project immediately. It would have been less disruptive if this library relied on a specific version of ash. (If needed, feature flags could be used to simultaneously support multiple versions of ash, i.e "ash-32", "ash-33", "ash-latest", etc.)

**Comment by Chris--B** [comment]
This issue is also broke projects I work on.

Cargo has surprisingly sophisticated version controls, so I added a maximum version check and have uploaded it here:
https://github.com/Chris--B/vk-mem-rs

I'm not sure if that's the right fix, so I have not made a PR yet.

**Comment by aclysma** [comment]
I made two separate PRs. #56 is a minimal ash 0.33 build fix, but would require all downstream users to update to ash 0.33.

Two approaches:

## Simplest possible fix
 * Make this change (https://github.com/Chris--B/vk-mem-rs/commit/89ad244d08fc82ed445e302830ec1202acd0b85b) and publish as 0.2.3
 * Then take #56 and publish as 0.3.0

This does not prevent future semver breaks.

## A more flexible approach
 * Take #55 but change the default feature to "ash_32", publish as 0.2.3
 * Then change the default to "ash_33" or "ash_latest" depending on whether the default should support semver and publish as 0.3.0

The second approach is slightly more complex and could become more complex if there are many variants of ash supported and the required changes between them are significant. Based on the current differences and the rate of change of the upstream ash crate, I think this could work pretty well. (I use this approach with a library that uses winit, and it's worked ok so far. The imgui crate also uses this approach.)

**Comment by aclysma** [comment]
Just realized there are some breaking changes like #36 so a republish of 0.2.2 probably should be just already published 0.2.2 + change the ash version as @Chris--B suggested

**Comment by aloucks** [comment]
Every ash releases breaks projects for me as well due to the semver-incompatible range specifier. Having a non-1.0 library in the public API of vk-mem is problematic. I think the "right" solution is to remove ash altogether and make vk-mem independent of the vulkan bindings (i.e. pass raw handles), but it's a non-trivial amount of work.

**Comment by repi** [comment]
@aclysma the second option definitely sounds like the most robust option. And yeah it can be somewhat messy if one does support lots of versions, though most of them would have identical `ash` code interaction still.

Unfortunately I think this crate is effectively unmaintained, we've moved to the pure Rust [`gpu-allocator`](https://github.com/Traverse-Research/gpu-allocator) instead

**Comment by MarijnS95** [comment]
> Every ash releases breaks projects for me as well due to the semver-incompatible range specifier. Having a non-1.0 library in the public API of vk-mem is problematic. I think the "right" solution is to remove ash altogether and make vk-mem independent of the vulkan bindings (i.e. pass raw handles), but it's a non-trivial amount of work.

@aloucks I don't think Ash is at fault here. Without a 1.0 release we're allowed to make _breaking_ changes by bumping the left-most non-zero number, in this case the minor version.

Not having an upper-bound in vk-mem however is problematic. Using default caret matching (allowing just `>=0.27.1, <0.28`) would have prevented this case, but I can see how it is equally annoying to make specific vk-mem versions just to support multiple versions of ash (since downstream crates using vk-mem would have to update at the same pace). That said, ash doesn't see that frequent updates anyway...

In the end the "right" solution is to have a 1.0 release of Ash, solidifying the API in the process. The problem with this autogenerated library is that we keep coming across ways to improve or simplify the user-facing bits. We'll also have to "hide" the experimental bits from Vulkan as API changes are expected there without us wanting to bump Ash.

**Comment by aloucks** [comment]
I wasn't blaming ash. I only meant that having a dependency in the public API that is still evolving is problematic. And yes, the range specifier is the biggest issue here. There isn't a way to pin the version of ash without `Cargo.lock` which makes libraries that depend on vk-mem a little frustrating.

**Comment by cheako** [comment]
Crates depending on vkmem can't publish currently?

**Comment by cheako** [comment]
I don't know why ppl suggest `gpu-allocator` as some kind of replacement...  Sure if you don't think Panasonic and Sony are different, then I guess.  The accolades of using AMD's memory allocator VS something maintained by a third party are clear?  I realized as I wrote that, that I don't know anything about `gpu-allocator`s owners.  Owned by AMD is not something the readme points too, so I could be forgiven.
